File: src/main/java/com/jcloisterzone/game/capability/FlowersCapability.java
package com.jcloisterzone.game.capability;

import com.jcloisterzone.Player;
import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.board.pointer.ScoreMeeplePositionsPointer;
import com.jcloisterzone.event.ExprItem;
import com.jcloisterzone.event.PointsExpression;
import com.jcloisterzone.event.ScoreEvent;
import com.jcloisterzone.event.ScoreEvent.ReceivedPoints;
import com.jcloisterzone.feature.City;
import com.jcloisterzone.feature.CompletableFeature;
import com.jcloisterzone.feature.Field;
import com.jcloisterzone.feature.RangeFeature;
import com.jcloisterzone.feature.Road;
import com.jcloisterzone.feature.Feature;
import com.jcloisterzone.feature.Monastery;
import com.jcloisterzone.feature.Scoreable;
import com.jcloisterzone.figure.Follower;
import com.jcloisterzone.figure.Meeple;
import com.jcloisterzone.figure.RangeFigure;
import com.jcloisterzone.figure.Special;
import com.jcloisterzone.game.Capability;
import com.jcloisterzone.game.capability.trait.FlowersBonusAffected;
import com.jcloisterzone.game.ScoreFeatureReducer;
import com.jcloisterzone.game.Token;
import com.jcloisterzone.game.state.GameState;
import com.jcloisterzone.game.state.PlacedTile;
import com.jcloisterzone.game.state.PlayersState;
import com.jcloisterzone.game.token.FlowersToken;
import com.jcloisterzone.random.RandomGenerator;
import com.jcloisterzone.reducers.AddPoints;
import io.vavr.collection.HashMap;
import io.vavr.collection.HashSet;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.collection.Set;
import io.vavr.collection.Stream;
import io.vavr.Tuple2;
import java.util.ArrayList;

public class FlowersCapability extends Capability<Void> {

    private static final long serialVersionUID = 1L;

    @Override
    public GameState onStartGame(GameState state, RandomGenerator random) {
        PlayersState ps = state.getPlayers();
        int multiplier = (int) Math.ceil(ps.getPlayers().size() / 4.0); // 1 for 1–4 players, 2 for 5–8, etc.

        List<FlowersToken> pool = List.empty();
        for (int i = 0; i < multiplier; i++) {
            pool = pool.appendAll(List.of(FlowersToken.values()));
        }

        for (Player p : ps.getPlayers()) {
            int idx = random.getNextInt(pool.size());
            FlowersToken token = pool.get(idx);
            pool = pool.removeAt(idx);
            ps = ps.setTokenCount(p.getIndex(), token, 1);
        }

        return state.setPlayers(ps);
    }

    @Override
    public List<ReceivedPoints> appendFiguresBonusPoints(GameState state, List<ReceivedPoints> bonusPoints, Scoreable feature, boolean isFinal) {
        Map<String,Integer> featureFlowers = HashMap.empty();

        Stream<PlacedTile> tiles = Stream.empty();
        if (feature instanceof Field field && isFinal) {
            featureFlowers = field.getModifier(state, Field.FLOWERS, HashMap.empty() );
        	tiles = field
    		            .getTilePositions()
    		            .map(pos -> state.getPlacedTiles().get(pos).get()).toStream();
        } else if (feature instanceof FlowersBonusAffected) {
            if (feature instanceof Monastery) {
                if (((Monastery) feature).isSpecialMonastery(state)) {
                    if (isFinal) {
                        tiles = ((Monastery) feature).getRangeTilesSpecialMonastery(state);
                    }
                } else {
                    tiles = ((Monastery) feature).getRangeTiles(state);
                }
            } else if (feature instanceof RangeFeature) {
                tiles = ((RangeFeature) feature).getRangeTiles(state);
            } else if (feature instanceof CompletableFeature){
            	tiles = ((CompletableFeature) feature)
            		        .getTilePositions()
            		        .map(pos -> state.getPlacedTiles().get(pos).get()).toStream();
            } else {
                throw new IllegalArgumentException("Unknown feature type for FlowersBonus " + feature);
            }
            featureFlowers = getFlowersOnTiles(state, tiles);
        }
        
        if (!featureFlowers.isEmpty()) {
            Stream<Follower> followers = (
                feature instanceof Monastery monastery && monastery.isSpecialMonastery(state)
                    ? Stream.ofAll(monastery.getMonasteryFollowers2(state)).map(Tuple2::_1)
                    : Stream.ofAll(feature.getFollowers(state))
                ).distinctBy(Follower::getPlayer);

            bonusPoints = appendBonusPoints(state, bonusPoints, featureFlowers, followers.map(Meeple.class::cast), tiles.map(t -> t.getPosition()).toSet());
        }
        return bonusPoints;
    }
    
    @Override
    public List<ReceivedPoints> appendSpecialFiguresBonusPoints(GameState state, List<ReceivedPoints> bonusPoints, Special figure, boolean isFinal) {
        Stream<PlacedTile> tiles = Stream.empty();
           if (figure instanceof FlowersBonusAffected) {
               tiles = ((RangeFigure) figure).getRangeTiles(state);
        }
        Map<String,Integer> flowersInRange = getFlowersOnTiles(state, tiles);
        bonusPoints = appendBonusPoints(state, bonusPoints, flowersInRange, Stream.of((Meeple) figure), tiles.map(PlacedTile::getPosition).toSet());
        return bonusPoints;
    }
    
    private Map<String,Integer> getFlowersOnTiles(GameState state, Stream<PlacedTile> tiles) {
        Map<String,Integer> flowers = HashMap.empty();
        for (PlacedTile t : tiles) {
            Map<FeaturePointer, Field> fields =
                t.getTile().getInitialFeatures()
                    .filterValues(f -> f instanceof Field)
                    .mapValues(f -> (Field) f);

            for (Tuple2<FeaturePointer, Field> fieldEntry : fields) {
                Map<String,Integer> fieldFlowers = fieldEntry._2.getModifier(state, Field.FLOWERS, HashMap.empty());
                flowers = flowers.merge(fieldFlowers, (a, b) -> a + b);
            }
        }
        return flowers;
    }
    
    private List<ReceivedPoints> appendBonusPoints(GameState state, List<ReceivedPoints> bonusPoints, Map<String,Integer> flowersList, Stream<Meeple> figures, Set<Position> positions) {
        if (figures.size()>0) {
//            Stream<Player> players = figures
//                    .flatMap(f -> f instanceof Follower ? Stream.of(((Follower) f).getPlayer())
//                               : f instanceof Special ? Stream.of(((Special) f).getPlayer())
//                               : Stream.empty());
//            Stream<Player> players = figures.map(Follower::getPlayer);
            PlayersState ps = state.getPlayers();
    
            HashMap<Token,Set<Player>> tokenPlayers = HashMap.empty();
            for(Token token: FlowersToken.values()) {
                tokenPlayers = tokenPlayers.put(token, ps.getPlayersWithToken(token));
            }
    
            for(Tuple2<String, Integer> flowers : flowersList) {
                Token token = FlowersToken.fromValue(flowers._1);
                for (Tuple2<Token,Set<Player>> p: tokenPlayers) {
                    ArrayList<ExprItem> exprItems = new ArrayList<ExprItem>();
                    if (token.equals(p._1)) {
                        ExprItem expr = new ExprItem(flowers._2, "flowers." + token.name(), 3 * flowers._2);
                        for (Player player: p._2) {
                            bonusPoints = bonusPoints.appendAll(
                                figures
                                    .filter(f -> f.getPlayer().equals(player))
                                    .map(f -> new ScoreEvent.ReceivedPoints(new PointsExpression("flowers", expr), player, new ScoreMeeplePositionsPointer(f.getDeployment(state), f.getId(), positions)))
                            );
//                            if (players.contains(player)) {
//                                bonusPoints = bonusPoints.append(new ScoreEvent.ReceivedPoints(new PointsExpression("flowers", expr), player, new MeeplePointer(fps.get(), positions)));
//                            }
                        }
                    }
                }
            }
        }
        return bonusPoints;
    }
}

File: src/main/ts/com/jcloisterzone/game/capability/FlowersCapability.ts
import { HashMap, type Map as VMap } from "../../../../io/vavr/Map.js";
import { List, Stream } from "../../../../io/vavr/SeqTypes.js";
import { HashSet, type Set } from "../../../../io/vavr/Set.js";
import type { Player } from "../../Player.js";
import type { Position } from "../../board/Position.js";
import { ScoreMeeplePositionsPointer } from "../../board/pointer/ScoreMeeplePositionsPointer.js";
import { ExprItem } from "../../event/ExprItem.js";
import { PointsExpression } from "../../event/PointsExpression.js";
import { ReceivedPoints } from "../../event/ScoreEvent.js";
import { CompletableFeature } from "../../feature/CompletableFeature.js";
import { Field } from "../../feature/Field.js";
import { Monastery } from "../../feature/Monastery.js";
import type { Scoreable } from "../../feature/Scoreable.js";
import { isInstanceOfRangeFeature } from "../../feature/RangeFeature.js";
import type { Follower } from "../../figure/Follower.js";
import type { Meeple } from "../../figure/Meeple.js";
import type { Special } from "../../figure/Special.js";
import type { RandomGenerator } from "../../random/RandomGenerator.js";
import { Capability } from "../Capability.js";
import { FlowersToken } from "../token/FlowersToken.js";
import { isInstanceOfFlowersBonusAffected } from "./trait/FlowersBonusAffected.js";
import type { GameState } from "../state/GameState.js";
import type { PlacedTile } from "../state/PlacedTile.js";

/** Flowers (Village Life): every player draws one random flower-colour token at game
 *  start (visible in player tokens); whenever a feature is scored, each follower owner
 *  whose token colour matches flowers on the feature's tiles gets +3 per flower —
 *  emitted as an extra "flowers" scoring event. */
export class FlowersCapability extends Capability<void> {
  override onStartGame(state: GameState, random: RandomGenerator): GameState {
    let ps = state.getPlayers();
    const multiplier = Math.ceil(ps.getPlayers().length() / 4.0); // 1 for 1–4 players, 2 for 5–8, …

    let pool: List<FlowersToken> = List.empty<FlowersToken>();
    for (let i = 0; i < multiplier; i++) {
      pool = pool.appendAll(FlowersToken.values()) as List<FlowersToken>;
    }

    for (const p of ps.getPlayers()) {
      const idx = random.getNextInt(pool.size());
      const token = pool.get(idx);
      pool = pool.removeAt(idx) as List<FlowersToken>;
      ps = ps.setTokenCount(p.getIndex(), token, 1);
    }

    return state.setPlayers(ps);
  }

  override appendFiguresBonusPoints(
    state: GameState,
    bonusPoints: List<ReceivedPoints>,
    feature: Scoreable,
    isFinal: boolean,
  ): List<ReceivedPoints> {
    let featureFlowers: VMap<string, number> = HashMap.empty<string, number>();
    let tiles: Stream<PlacedTile> = Stream.empty<PlacedTile>();

    if (feature instanceof Field && isFinal) {
      featureFlowers = feature.getModifier(state, Field.FLOWERS, HashMap.empty<string, number>());
      tiles = Stream.ofAll(feature.getTilePositions()).map(
        (pos) => state.getPlacedTiles().get(pos).get(),
      ) as Stream<PlacedTile>;
    } else if (isInstanceOfFlowersBonusAffected(feature)) {
      if (feature instanceof Monastery) {
        if (feature.isSpecialMonastery(state)) {
          if (isFinal) {
            tiles = feature.getRangeTiles(state);
          }
        } else {
          tiles = feature.getRangeTiles(state);
        }
      } else if (isInstanceOfRangeFeature(feature)) {
        tiles = (feature as unknown as { getRangeTiles(s: GameState): Stream<PlacedTile> }).getRangeTiles(state);
      } else if (feature instanceof CompletableFeature) {
        tiles = Stream.ofAll(feature.getTilePositions()).map(
          (pos) => state.getPlacedTiles().get(pos).get(),
        ) as Stream<PlacedTile>;
      } else {
        throw new Error("Unknown feature type for FlowersBonus " + String(feature));
      }
      featureFlowers = this.getFlowersOnTiles(state, tiles);
    }

    if (!featureFlowers.isEmpty()) {
      const followers: Stream<Follower> = (
        feature instanceof Monastery && feature.isSpecialMonastery(state)
          ? (Stream.ofAll(feature.getMonasteryFollowers2(state)).map((t) => t._1) as Stream<Follower>)
          : (feature.getFollowers(state) as Stream<Follower>)
      ).distinctBy((f: Follower) => f.getPlayer()) as Stream<Follower>;

      let positions: Set<Position> = HashSet.empty<Position>();
      for (const t of tiles) positions = positions.add(t.getPosition());
      bonusPoints = this.appendBonusPoints(
        state,
        bonusPoints,
        featureFlowers,
        followers as unknown as Stream<Meeple>,
        positions,
      );
    }
    return bonusPoints;
  }

  override appendSpecialFiguresBonusPoints(
    state: GameState,
    bonusPoints: List<ReceivedPoints>,
    figure: Special,
    isFinal: boolean,
  ): List<ReceivedPoints> {
    void isFinal;
    let tiles: Stream<PlacedTile> = Stream.empty<PlacedTile>();
    if (isInstanceOfFlowersBonusAffected(figure)) {
      tiles = (figure as unknown as { getRangeTiles(s: GameState): Stream<PlacedTile> }).getRangeTiles(state);
    }
    const flowersInRange = this.getFlowersOnTiles(state, tiles);
    let positions: Set<Position> = HashSet.empty<Position>();
    for (const t of tiles) positions = positions.add(t.getPosition());
    return this.appendBonusPoints(
      state,
      bonusPoints,
      flowersInRange,
      Stream.of(figure as unknown as Meeple),
      positions,
    );
  }

  private getFlowersOnTiles(state: GameState, tiles: Stream<PlacedTile>): VMap<string, number> {
    let flowers: VMap<string, number> = HashMap.empty<string, number>();
    for (const t of tiles) {
      for (const entry of t.getTile().getInitialFeatures()) {
        if (!(entry._2 instanceof Field)) continue;
        const fieldFlowers = entry._2.getModifier(state, Field.FLOWERS, HashMap.empty<string, number>());
        for (const ff of fieldFlowers) {
          flowers = flowers.put(ff._1, flowers.get(ff._1).getOrElse(0) + ff._2);
        }
      }
    }
    return flowers;
  }

  private appendBonusPoints(
    state: GameState,
    bonusPoints: List<ReceivedPoints>,
    flowersList: VMap<string, number>,
    figures: Stream<Meeple>,
    positions: Set<Position>,
  ): List<ReceivedPoints> {
    if (figures.size() > 0) {
      const ps = state.getPlayers();
      const playersWithToken = (token: FlowersToken): Player[] =>
        ps
          .getPlayers()
          .toArray()
          .filter((p) => ps.getPlayerTokenCount(p.getIndex(), token) > 0);

      for (const flowers of flowersList) {
        const token = FlowersToken.fromValue(flowers._1);
        const expr = new ExprItem(flowers._2, "flowers." + token.name(), 3 * flowers._2);
        for (const player of playersWithToken(token)) {
          bonusPoints = bonusPoints.appendAll(
            figures
              .filter((f) => f.getPlayer().equals(player))
              .map(
                (f) =>
                  new ReceivedPoints(
                    new PointsExpression("flowers", expr),
                    player,
                    new ScoreMeeplePositionsPointer(f.getDeployment(state)!, f.getId(), positions),
                  ),
              ),
          ) as List<ReceivedPoints>;
        }
      }
    }
    return bonusPoints;
  }
}

Capability.register(FlowersCapability);
